crates/jstz_cli/src/repl.rs

```diff
@@ -5,41 +5,97 @@ use jstz_api::{
     encoding::EncodingApi, http::HttpApi, js_log::set_js_logger, url::UrlApi,
     urlpattern::UrlPatternApi, ConsoleApi, KvApi,
 };
-use jstz_core::host::HostRuntime;
 use jstz_core::{
+    host::HostRuntime,
     host_defined,
     kv::Kv,
     runtime::{self, Runtime},
 };
 use jstz_proto::api::{ContractApi, LedgerApi};
-use rustyline::{error::ReadlineError, Editor};
+use rustyline::{
+    completion::Completer, error::ReadlineError, highlight::Highlighter, hint::Hinter,
+    validate::Validator, Editor, Helper,
+};
+use std::borrow::Cow;
 use tezos_smart_rollup_mock::MockHost;
 
 use syntect::{
     easy::HighlightLines,
     highlighting::{Style, Theme, ThemeSet},
     parsing::{SyntaxReference, SyntaxSet},
-    util::LinesWithEndings,
-};
-
-use crossterm::{
-    cursor::MoveUp,
-    terminal::{Clear, ClearType},
-    ExecutableCommand,
 };
-use std::io::{stdout, Write};
 
 use crate::{config::Config, debug_api::DebugApi};
 
-pub fn exec(self_address: Option<String>, cfg: &Config) -> Result<()> {
-    // Initialize syntax set and theme set
-    let ss = SyntaxSet::load_defaults_newlines();
-    let ts = ThemeSet::load_defaults();
+struct JsHighlighter {
+    ss: SyntaxSet,
+    syntax: SyntaxReference,
+    theme: Theme,
+}
+
+impl JsHighlighter {
+    pub fn new() -> Self {
+        // Initialize syntax set and theme set only once
+        let ss = SyntaxSet::load_defaults_newlines();
+        let ts = ThemeSet::load_defaults();
+
+        // Find JavaScript syntax and a theme
+        let syntax = ss.find_syntax_by_extension("js").unwrap();
+        let theme = &ts.themes["base16-ocean.dark"];
+
+        JsHighlighter {
+            ss: ss.clone(),
+            syntax: syntax.clone(),
+            theme: theme.clone(),
+        }
+    }
+
+    fn apply_foreground_only(&self, styles: &[(Style, &str)]) -> String {
+        let mut result = String::new();
 
-    // Find JavaScript syntax and a theme (e.g., base16-ocean.dark)
-    let syntax = ss.find_syntax_by_extension("js").unwrap();
-    let theme = &ts.themes["base16-ocean.dark"];
+        for &(style, text) in styles {
+            let color = style.foreground;
+            let color_string = format!("\x1b[38;2;{};{};{}m", color.r, color.g, color.b);
 
+            result.push_str(&color_string);
+            result.push_str(text);
+
+            result.push_str("\x1b[0m");
+        }
+
+        result
```

Comment:
Slightly more Rusty version:

```rust
        styles
            .iter()
            .map(|&(style, text)| {
                let color = style.foreground;
                format!(
                    "\x1b[38;2;{};{};{}m{}\x1b[0m",
                    color.r, color.g, color.b, text
                )
            })
            .collect()
 ```
 
 (GitHub won't let me propose this as a suggestion due to the deleted lines in the diff)

Reply:
This exact suggestion doesn't pass lint. It needs fold instead of map.